maitake/src/wait/map.rs

```diff
@@ -0,0 +1,926 @@
+use crate::{
+    loom::{
+        cell::UnsafeCell,
+        sync::{
+            atomic::{AtomicUsize, Ordering::*},
+            spin::{Mutex, MutexGuard},
+        },
+    },
+    util,
+};
+use cordyceps::{
+    list::{self, List},
+    Linked,
+};
+use core::{
+    fmt::Debug,
+    future::Future,
+    marker::PhantomPinned,
+    mem::{self, MaybeUninit},
+    pin::Pin,
+    ptr::NonNull,
+    task::{Context, Poll, Waker},
+};
+use mycelium_bitfield::FromBits;
+use mycelium_util::fmt;
+use mycelium_util::sync::CachePadded;
+use pin_project::{pin_project, pinned_drop};
+
+#[cfg(test)]
+mod tests;
+
+/// An error indicating a failed wake
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum WaitError {
+    Closed,
+    AlreadyConsumed,
+    NeverAdded,
+}
+
+/// An indication of the result of a call to `wait()`
+// Note: We duplicate the WaitResult type as it currently only has "Closed"
+// as an error condition. We should merge these two on the next breaking
+// change.
```

Comment:
since this crate has never been published, it's fine to do that now --- although, honestly, i think it's fine (and perhaps better) for the `WaitMap` type to have a different error type than the other types in the wait module. two of those error conditions can never occur with `WaitQueue` or `WaitCell`, so i have a slight preference for them returning an error type where users will not have to match the error and handle two cases that will never happen.

Reply:
Makes sense! Will remove the comment.